src/mgmt/srv_system.c

```diff
@@ -596,6 +632,133 @@ ds_mgmt_svc_put_leader(struct mgmt_svc *svc)
+	struct server_rec	*rec;
+	struct server_entry	entry = {};
+	d_iov_t			key;
+	d_iov_t			value;
+	uint32_t		map_version;
+	int			rc;
+
+	rc = rdb_tx_begin(svc->ms_rsvc.s_db, svc->ms_rsvc.s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_put, rc);
+
+	ABT_rwlock_wrlock(svc->ms_lock);
+	d_iov_set(&key, &rank, sizeof(rank));
+	d_iov_set(&value, NULL, sizeof(*rec));
+	rc = rdb_tx_lookup(&tx, &svc->ms_servers, &key, &value);
+	if (rc != 0) {
+		D_DEBUG(DB_MGMT, "Can not find server rank %u record: %d\n",
+			rank, rc);
+		D_GOTO(unlock, rc);
+	}
+
+	rec = value.iov_buf;
+	if (!(rec->sr_flags & SERVER_IN)) {
+		D_DEBUG(DB_MGMT, "rank %u already excluded\n", rank);
+		D_GOTO(unlock, rc = 0);
+	}
+
+	/* Update ms_server for the rank */
+	rec->sr_flags &= ~SERVER_IN;
+	d_iov_set(&value, rec, sizeof(*rec));
+	rc = rdb_tx_update(&tx, &svc->ms_servers, &key, &value);
+	if (rc != 0)
+		D_GOTO(unlock, rc);
+
+	/* Update the map version */
+	D_DEBUG(DB_MGMT, "upgrade version %u->%u\n", svc->ms_map_version,
+		svc->ms_map_version + 1);
+	map_version = svc->ms_map_version + 1;
+	d_iov_set(&value, &map_version, sizeof(map_version));
+	rc = rdb_tx_update(&tx, &svc->ms_root, &ds_mgmt_prop_map_version,
+			   &value);
+	if (rc != 0) {
+		D_ERROR("failed to increment map version to %u: %d\n",
+			map_version, rc);
+		D_GOTO(unlock, rc);
+	}
+
+	rc = rdb_tx_commit(&tx);
+	if (rc != 0) {
+		D_ERROR("failed to commit map version %u: %d\n", map_version,
+			rc);
+		D_GOTO(unlock, rc);
+	}
+
+	svc->ms_map_version = map_version;
+	entry.se_rank = rank;
+	rc = ds_mgmt_group_update(CRT_GROUP_MOD_OP_REMOVE, &entry,
+				  1 /* nservers */, map_version);
+	if (rc != 0) {
+		rdb_resign(svc->ms_rsvc.s_db, svc->ms_rsvc.s_term);
+		D_GOTO(unlock, rc = 0);
+	}
+
+	ds_rsvc_request_map_dist(&svc->ms_rsvc);
+
+unlock:
+	ABT_rwlock_unlock(svc->ms_lock);
+	rdb_tx_end(&tx);
+out_put:
+	ds_mgmt_svc_put_leader(svc);
+	D_FREE_PTR(arg);
+}
+
+void
+ds_mgmt_crt_event_cb(d_rank_t rank, enum crt_event_source src,
+		     enum crt_event_type type, void *arg)
+{
+	struct ds_mgmt_evict_arg	*ult_arg;
+	struct mgmt_svc			*svc;
+	int				rc = 0;
+
+	/* Only used for evict the rank for the moment */
+	if (src != CRT_EVS_SWIM || type != CRT_EVT_DEAD) {
+		D_DEBUG(DB_MGMT, "ignore src/type %u/%u\n", src, type);
+		return;
+	}
+
+	rc = ds_mgmt_svc_lookup_leader(&svc, NULL);
+	if (rc != 0) {
+		D_ERROR("Can not lookup leader: rc %d\n", rc);
+		return;
+	}
+
+	D_ALLOC_PTR(ult_arg);
+	if (ult_arg == NULL) {
+		ds_mgmt_svc_put_leader(svc);
+		D_ERROR("ult arg allocation failed\n");
+		return;
+	}
+
+	/* Since this is called from swim_progress, let's create another
+	 * ult to do the job asynchronously, so this can return quickly.
+	 */
+	ult_arg->svc = svc;
+	ult_arg->rank = rank;
+	rc = dss_ult_create(ds_mgmt_evict_rank_ult, ult_arg, DSS_ULT_MISC,
+			    DSS_TGT_SELF, 0, NULL);
+	if (rc) {
+		D_ERROR("create evict ult failed: rc %d\n", rc);
+		ds_mgmt_svc_put_leader(svc);
+		D_FREE_PTR(ult_arg);
```

Comment:
A minor suggestion: insert a return statement here (just in case in the future any more statements are added before this function returns - in which case it would be easy for the author of that new code to not notice the lack of return here).

Reply:
ok